src/UUPSProxy.sol

```diff
@@ -60,5 +74,42 @@ contract UUPSProxy is Proxy, IUUPSProxy {
         _SALT_SLOT.erc7201Slot().getBytes32Slot().value = _salt;
     }
 
+    function _delegate(address implementation, bool checkImplementation) internal {
+        bytes32 implementationSlot = ERC1967Utils.IMPLEMENTATION_SLOT;
+        assembly {
+            // Copy msg.data. We take full control of memory in this inline assembly
+            // block because it will not return to Solidity code. We overwrite the
+            // Solidity scratch pad at memory position 0.
+            calldatacopy(0, 0, calldatasize())
+
+            // Call the implementation.
+            // out and outsize are 0 because we don't know the size yet.
+            let result := delegatecall(gas(), implementation, 0, calldatasize(), 0, 0)
+
+            // check if implementation has changed
+            let implAfter := sload(implementationSlot)
+
+            if and(checkImplementation, iszero(eq(implAfter, implementation))) {
+                revert(0, 0)
```

Comment:
It's awkward, but it'd be better if we can revert with an actual error code. You can use `errortype.selector` (outside assembly) to get the 4-byte selector to revert with.